### PR TITLE
feat(ag-trade): fetch, (un)marshal, lookup, sign... add to reserve

### LINKED ISSUE #1: caplets for trading with the Agoric smartWallet
I aim to migrate them to this repo from

 - https://github.com/dckc/finquick/pull/64

See also:

 - https://github.com/Agoric/agoric-sdk/discussions/8551

cc @kriskowal @tgrecojs

### PR DESCRIPTION
**Pull request by dckc**
fixes #1 

After setting up lint/test devtools and such (#6), I'm migrating stuff from finquick and swaparoo.

A hello-world chain interaction works:

```
~/projects/playground/packages/ag-trade$ make clean
endo reset

~/projects/playground/packages/ag-trade$ make node-info
++ install cosmos fetch plugin
endo make --UNSAFE src/cosmosFetch.js -n cosmos-fetch
Object [Alleged: CosmosFetch] {}

endo make src/net-local.js -n local -p cosmos-fetch
{ lcd: Object [Alleged: LCD] {}, rpc: Object [Alleged: RpcClient] {} }

endo eval "E(local.lcd).getJSON('/node_info').then(i => i.node_info)" local
{
  channels: '40202122233038606100',
  id: '8da5e462e560eb951e2c6d8ad2f9ce50c58a0be6',
  listen_addr: 'tcp://0.0.0.0:26656',
  moniker: 'localnet',
  network: 'agoriclocal',
  other: { rpc_address: 'tcp://0.0.0.0:26657', tx_index: 'on' },
  protocol_version: { app: '0', block: '11', p2p: '8' },
  version: '0.34.23'
}
```

cc @tgrecojs @kriskowal

### CODE REVIEW

**Comment by tgrecojs** [comment]
great hello, world @dckc!

I had to remove `shell` from line 1 so that it would work on MacOS (shown below)

```
STATE=$(endo where state)
``` 

This took a whopping ~5 seconds.... *aaand voila!* 🪄 

ss_01102024_000093

🚀 amazing way to break ground on (what i expect to be the) safest adult playground to ever grace the interwebs 😎

**Comment by dckc** [comment]
## smart wallet offers such as add to reserve

Prerequisite: `ALICE_SECRET` env has the 24 words corresponding to an account with a smart wallet already provisioned a la [agoric-3-proposals](https://github.com/Agoric/agoric-3-proposals/) as used in [Your First Agoric Dapp](https://docs.agoric.com/guides/getting-started/)

```
$ direnv allow
direnv: loading ~/projects/playground/.envrc
direnv: export +ALICE_SECRET +JACK_SECRET ~PATH
```

Then:

```
~/projects/playground/packages/ag-trade$ make clean
endo reset

~/projects/playground/packages/ag-trade$ make reserve-add
++ install cosmos fetch plugin
endo make --UNSAFE src/cosmosFetch.js -n cosmos-fetch
Object [Alleged: CosmosFetch] {}

endo make src/net-local.js -n local -p cosmos-fetch
{ lcd: Object [Alleged: LCD] {}, rpc: Object [Alleged: RpcClient] {} }

++ start smart wallet caplet -- TODO: should not need UNSAFE
endo make --UNSAFE src/smartWallet.js -n client-maker
Object [Alleged: SmartWalletFactory] {}
endo eval "E(wf).makeWalletKit('survey thank ...', local.rpc, local.lcd)" \
        local wf:client-maker -n alice-wk
{
  query: Object [Alleged: QueryTool] {},
  smartWallet: Object [Alleged: SmartWallet] {},
  tx: Object [Alleged: SigningClient] {}
}

endo make --UNSAFE src/fresh-id.js -n fresh
Object [Alleged: FreshMaker] {}

endo eval "E(fresh).start('trade-')" fresh -n fresh-trade-id
Object [Alleged: Fresh] {}

endo mkguest alice
Object [Alleged: EndoGuest] {}

endo send alice @wallet:alice-wk
endo send alice @fresh:fresh-trade-id
endo inbox --as alice
0. "HOST" sent "@wallet" at "2024-01-10T07:00:27.268Z"
1. "HOST" sent "@fresh" at "2024-01-10T07:00:27.362Z"
# these 0/1 message numbers are fragile
endo adopt 0 wallet --as alice
endo adopt 1 fresh --as alice

endo run src/reserve-add.js 1 IST -p alice
{
  status: {
    id: 'trade-0700.0',
    invitationSpec: {
      callPipe: [Array],
      instancePath: [Array],
      source: 'agoricContract'
    },
    numWantsSatisfied: 1,
    payouts: { Collateral: [Object] },
    proposal: { give: [Object] },
    result: 'added Collateral to the Reserve'
  },
  tx: {
    height: 26737,
    transactionHash: '937A033D429EF5E00774C2A94109E7627C784115BF0EB57BA4C9B91E86D6D5CA'
  }
}
```

File: packages/ag-trade/src/smartWallet.js [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+ * @file Agoric smart wallet proxy
+ */
+// @ts-check
+
+import { E, Far } from '@endo/far';
+import {
+  SigningStargateClient,
+  assertIsDeliverTxSuccess,
+} from '@cosmjs/stargate';
+import { Tendermint34Client } from '@cosmjs/tendermint-rpc';
```
Comment (dckc) [inline packages/ag-trade/src/smartWallet.js:11]: This shouldn't have to be an `--UNSAFE` plugin -- the network stuff is all in a separate vat -- but I'm struggling to get protobuf stuff to run confined. See:

 - https://github.com/endojs/endo/discussions/1875

File: packages/ag-trade/src/hdWallet.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+  /** @type {import('@cosmjs/proto-signing').OfflineDirectSigner} */
+  const methods = {
+    getAccounts: () => wallet.getAccounts(),
+    signDirect: (address, signDoc) => wallet.signDirect(address, signDoc),
+  };
+  return Far('HdWallet', methods);
+};
+
+export const make = async () => Far('HdWalletMaker', { fromMnemonic });
```
Comment (dckc) [inline packages/ag-trade/src/hdWallet.js:34]: computing signatures should be able to run in a separate caplet from the rest of the smartWallet code, but we can't marshal byte sequences:

 - https://github.com/endojs/endo/issues/1331#issuecomment-1819591785

File: packages/ag-trade/src/marshalTables.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  };
+
+  return harden({ convertValToSlot, convertSlotToVal });
+};
+
+/** @type {(slot: unknown, iface: string | undefined) => any} */
+const synthesizeRemotable = (slot, iface) =>
+  Far(`${(iface ?? '').replace(/^Alleged: /, '')}#${slot}`, {});
+
+export const makeClientMarshaller = () => {
```
Comment (dckc) [inline packages/ag-trade/src/marshalTables.js:41]: This should perhaps be imported from ui-kit, but the version there has an issue:

 - https://github.com/Agoric/ui-kit/issues/73

**Comment by dckc** [comment]
## Goal: PetName interop

While working on some prototypes Agoric dapps in an internal hackathon, we had `agoric1...` addresses on the screen here and there. Even form fields where the user entered and address. I thought to myself:

 1. There's got to be a better way. We should be using petnames.
 2. Spritely has some petname stuff. I wonder if it helps?
 3. Using Spritely stuff directly in Agoric apps seems like a stretch.
 4. In OCapN discussions, we've talked about an [CapTP bridge](https://github.com/ocapn/ocapn/issues/6#issuecomment-1477952429) between Spritely/OCapN stuff and endo.
 5. I've done endo plug-ins for Google Sheets... why not the Agoric chain?

## Sub-goal: ping Agoric chain using a `fetch` endo plugin

Get the `endo` CLI in your path. It's on the `endo` branch (currently 2023-12-30 12:27 096879a37). I use direnv to manage `$PATH`; it's also handy for managing some secrets we'll use later.

```
$ cd projects/playground/packages/ag-trade/
direnv: loading ~/projects/playground/.envrc
direnv: export +ALICE_SECRET +JACK_SECRET ~PATH

$ endo where state
/home/connolly/.local/state/endo
```

The default `Makefile` target in `ag-trade` also does `endo where state`:

```
$ make
state: /home/connolly/.local/state/endo
```

Start an agoric chain using `docker compose` and `agoric-3-proposals`. _IOU details._

We run `cosmosFetch.js` unconfined to make an object that can make net access endpoints used in network protocols of blockchains based on Cosmos SDK:
  1. Make a JSON RPC endpoint given an URL.
  2. Make an endpoint that, given a path, does a GET and the JSON.parse of the content.

```
$ make fetch-plug-in
++ install cosmos fetch plugin
endo make --UNSAFE src/cosmosFetch.js -n cosmos-fetch
Object [Alleged: CosmosFetch] {}
```

Now we can run `net-local.js` unconfined, giving it `cosmos-fetch` to get a pair of endpoints, and then use `endo eval` to use one of those endpoints to get some info from a local node:

```
$ make node-info
endo make src/net-local.js -n local -p cosmos-fetch
{ lcd: Object [Alleged: LCD] {}, rpc: Object [Alleged: RpcClient] {} }
endo eval "E(local.lcd).getJSON('/node_info').then(i => i.node_info)" local
{
  channels: '40202122233038606100',
  id: '8da5e462e560eb951e2c6d8ad2f9ce50c58a0be6',
  listen_addr: 'tcp://0.0.0.0:26656',
  moniker: 'localnet',
  network: 'agoriclocal',
  other: { rpc_address: 'tcp://0.0.0.0:26657', tx_index: 'on' },
  protocol_version: { app: '0', block: '11', p2p: '8' },
  version: '0.34.23'
}
```

**Pull request merged**